src/hotspot/share/opto/superword.cpp

```diff
@@ -2877,10 +2883,139 @@ bool SuperWord::output() {
   if (do_reserve_copy()) {
     make_reversable.use_new();
   }
+
+  for (uint i = 0; i < unordered_reductions.size(); i++) {
+    UnorderedReductionNode* ur = unordered_reductions.at(i)->as_UnorderedReduction();
+    move_unordered_reduction_out_of_loop(ur);
+  }
```

Comment:
Hi @eme64 , if we move this processing post SLP to a stand alone pass, we can also handler vector IR created through VectorAPI.

Reply:
We can also relax following limitation with your patch since loop body will now comprise of lane wise vector operations with reduction moved out of loop it may allow vectorizing  patterns like  res += a[i];  which is composed of single load and reduction operation.
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/superword.cpp#L2265

Reply:
@jatin-bhateja 

> if we move this processing post SLP to a stand alone pass, we can also handler vector IR created through VectorAPI.

Where exactly would you put it? We need a location during LoopOpts, so that we have the ctrl information. I previously suggested in `split_if`, but @vnkozlov seemed not very excited. Additionally, I have not seen any case where VectorAPI could make use of it. I gave it a quick look, so maybe you can find something.

And about:
https://github.com/openjdk/jdk/blob/941a7ac7dab243c6033a78880fd31faa803e62ab/src/hotspot/share/opto/superword.cpp#L2265

Yes, that is the hope, that we could allow things like that to vectorize. The question is if we can guarantee that my new optimization will happen. But probably it is ok to be a bit optimistic here.

Reply:
There is also an opportunity to support reduction involving non-commutative bytecodes like isub and lsub, but it may need explicit backend support and can be taken up separately.